### PR TITLE
vcmi: migrate to C++17

### PR DESCRIPTION
**Pull request by rilian-la-te**
1. use std::any
2. use std::variant
3. use std::optional

### CODE REVIEW

File: Global.h [hunk 1]
```diff
@@ -576,24 +577,24 @@ namespace vstd
 		return i >= 0  &&  i < c.size();
 	}
 
-	template <typename Container, typename Index>
-	boost::optional<typename Container::const_reference> tryAt(const Container &c, Index i)
+	template<typename Container, typename Index>
+	std::optional<typename Container::value_type> tryAt(const Container & c, Index i)
```
Comment (IvanSavenko) [inline Global.h:581]: Why const_ref has been changed to value_type and only here?
Similar methods don't have similar change

File: lib/CArtHandler.cpp [hunk 2]
```diff
@@ -740,15 +740,16 @@ void CArtHandler::erasePickedArt(const ArtifactID & id)
 {
 	CArtifact *art = objects[id];
 
-	if(auto artifactList = listFromClass(art->aClass))
+	if(auto artl = listFromClass(art->aClass))
```
Comment (IvanSavenko) [inline lib/CArtHandler.cpp:743]: Why the change? Old code was fine with access via dereferences

File: lib/CArtHandler.cpp [hunk 3]
```diff
@@ -758,20 +759,20 @@ void CArtHandler::erasePickedArt(const ArtifactID & id)
 		logMod->warn("Problem: cannot find list for artifact %s, strange class. (special?)", art->getNameTranslated());
 }
 
-boost::optional<std::vector<CArtifact*>&> CArtHandler::listFromClass( CArtifact::EartClass artifactClass )
+std::optional<std::reference_wrapper<std::vector<CArtifact *>>> CArtHandler::listFromClass(CArtifact::EartClass artifactClass)
```
Comment (IvanSavenko) [inline lib/CArtHandler.cpp:762]: Perhaps remove this method and merge this code into `CArtHandler::erasePickedArt`?
This is the only one usage of this method and removal of optionals might actually clear up this code which got even more complex with this change

File: lib/serializer/BinaryDeserializer.h [hunk 4]
```diff
@@ -43,10 +44,21 @@ class DLL_LINKAGE BinaryDeserializer : public CLoaderBase
...
+		};
+
+		template <class V>
+		using mpl_types = typename mpl_types_impl<V>::type;
+
 		VariantLoaderHelper(Source & source):
 			source(source)
 		{
-			boost::mpl::for_each<typename Variant::types>(std::ref(*this));
+			boost::mpl::for_each<mpl_types<Variant>>(std::ref(*this));
```
Comment (IvanSavenko) [inline lib/serializer/BinaryDeserializer.h:61]: Love removing boost?
Perhaps we can replace this class with something like this:
https://stackoverflow.com/a/57643845
(and remove final usage of mpl along the way)

File: Global.h (hunk 1 above)
Comment (rilian-la-te) [reply Global.h:581]: @IvanSavenko because other similar methods are unused. Const_ref is not allowed in std::optional, and I tried but unsuccessful in using std::reference_wrapper here. Perhaps remove unused methods?

File: lib/CArtHandler.cpp (hunk 2 above)
Comment (rilian-la-te) [reply lib/CArtHandler.cpp:743]: @IvanSavenko because there is std::reference_wrapper now here. We should get a reference from it somewhere. So, it is split to 2 methods - firstly, get an optional, and second - get actual reference from wrapper, so, I actually split this check to 2.

File: Global.h (hunk 1 above)
Comment (IvanSavenko) [reply Global.h:581]: In that case:
- remove unused methods
- optionally - remove this method (only 1 usage in code)

File: Global.h [hunk 5]
```diff
@@ -101,6 +101,7 @@ static_assert(sizeof(bool) == 1, "Bool needs to be 1 byte in size.");
 #define _USE_MATH_DEFINES
 
 #include <algorithm>
+#include <any>
```
[the remark this answers is not in the document]
Comment (rilian-la-te) [reply Global.h:104]: done

File: Global.h (hunk 1 above)
Comment (rilian-la-te) [reply Global.h]: unused functions removed, method converted to lambda and moved inside VCAI

File: lib/CArtHandler.cpp (hunk 3 above)
Comment (rilian-la-te) [reply lib/CArtHandler.cpp]: done

File: lib/serializer/BinaryDeserializer.h (hunk 4 above)
Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.h:61]: About MPL - i think it is next time, because it is not a final usage.

Comment (rilian-la-te) [reply lib/serializer/BinaryDeserializer.h:61]: This will require a separate PR, not simple one as this.

**Comment by rilian-la-te** [comment]
Will fix CI tomorrow.

**Comment by rilian-la-te** [comment]
@IvanSavenko CI fixed, please, look again)

**Review by IvanSavenko — APPROVED** [review]
